### PR TITLE
Implement Move To Trash feature

### PR DESCRIPTION
**Pull request by oneeyeman1**
This is a second attempt to push "Move to Trash" feature to the library.

I made this function static and I hope I fixed all the documentation things.

Please review.

Next will be an asynchronous implementation of this feature.

Thank you and sorry for such a long delay.

### CODE REVIEW

File: include/wx/desktopenv.h
Comment (PBfordev) [inline include/wx/desktopenv.h:7]:
```suggestion
// Licence:     wxWindows licence
```

Applies everywhere.

File: include/wx/msw/desktopenv.h
Comment (PBfordev) [inline include/wx/msw/desktopenv.h:21]: Remove the unused access specifier.

File: interface/wx/desktopenv.h
Comment (PBfordev) [inline interface/wx/desktopenv.h:17]:
```suggestion
@since 3.1.6
```

File: interface/wx/desktopenv.h
Comment (PBfordev) [inline interface/wx/desktopenv.h:18]:
```suggestion
```

File: interface/wx/desktopenv.h
Comment (PBfordev) [inline interface/wx/desktopenv.h:24]:
```suggestion
    Moves the file/folder to the Recycle Bin/Trash. The path should be absolute.
```

The implementation makes the path absolute.

File: src/common/desktopenvcmn.cpp
Comment (PBfordev) [inline src/common/desktopenvcmn.cpp:21]: BCC is not supported anymore, its stuff needs to be removed everywhere.

File: src/common/desktopenvcmn.cpp
Comment (PBfordev) [inline src/common/desktopenvcmn.cpp:31]:
```suggestion
bool wxDesktopEnvBase::MoveToRecycleBin(const wxString& path)
```
Applies everywhere.

File: src/common/desktopenvcmn.cpp
Comment (PBfordev) [inline src/common/desktopenvcmn.cpp:39]:
```suggestion
        wxLogSysError( _( "Failed to move '%s' to Recycle Bin" ), path);
```

Additionally, I am not sure if `wxLogSysError()` is the best fit here. Do we always have last error set here?

File: src/common/desktopenvcmn.cpp
Comment (PBfordev) [inline src/common/desktopenvcmn.cpp:35]: I would prefer separating variable declarations from the rest of the code by an empty line.

File: src/msw/desktopenv.cpp
Comment (PBfordev) [inline src/msw/desktopenv.cpp:49]: `wxLogSysError()` should not be used here: `SHFileOperation()` does not affect the last error code.

Either just display the simple error message or if you want to display the specific error, you could do something like this
```cpp
wxLogError(_( "Moving '%s' to Recycle Bin failed with error 0x%08lx (%s)."),
    path, (long)res, wxSysErrorMsgStr(res));
```

**Review by PBfordev — COMMENTED** [review]
I still do not understand the API, i.e., why have a class with a single, static, method. I guess this is supposed to be extended in the future. 

Even if so, why not submit it complete so we have a big picture available?

Nevertheless, I added a bunch of (mostly) nitpick comments.

File: tests/file/desktopenv.cpp
Comment (PBfordev) [inline tests/file/desktopenv.cpp:30]:
```suggestion
#if defined(__WXMSW__) || defined(__WXGTK__) || ( defined(__WXOSX__) && !( wxOSX_USE_IPHONE ) )
```

File: interface/wx/desktopenv.h
Comment (PBfordev) [inline interface/wx/desktopenv.h:26]: I think return values are usually described in wxWidgets documentation like this
```suggestion
    @return @true if the file/folder was successfully moved to the Recycle Bin/Trash.
```

**Comment by oneeyeman1** [comment]
@PBfordev ,
Thank you for your input.
For the API - I plan to add asynchronous functions and possibly Restore From Trash/Recycle Bin.

And maybe I can add the wrapper for this on the global level - whatever Vadim will prefer.

Now - I'm also curious - where/how do I fix MSW cross-compile error? Add the missing symbol to msw/private.h?

File: include/wx/desktopenv.h
Comment (oneeyeman1) [reply include/wx/desktopenv.h]: @PBfordev ,
I fixed those, however, there are 2 more files with mentioning of "wxWidgets license" - mimetype.h/cpp.

I don't know what the plan is for them to be.

Vadim?

**Comment by oneeyeman1** [comment]
And I have a question:

Would it make sense to have as a public function:

static bool FilesInTrahs(const wxVector<wxString> &);

?

TIA!!

File: include/wx/desktopenv.h
Comment (vadz) [inline include/wx/desktopenv.h:31]: This should be just removed, as well as the corresponding header files themselves.

File: include/wx/desktopenv.h
Comment (vadz) [inline include/wx/desktopenv.h:15]: What's the point of having a "Base" class when there are no virtual functions here? This should be just `wxDesktopEnv` itself, it's a namespace-like class, it doesn't make any sense to inherit from it.

File: interface/wx/desktopenv.h
Comment (vadz) [inline interface/wx/desktopenv.h:15]: There is no real need to mention that a function is sync, they all are by default.

File: src/common/desktopenvcmn.cpp
Comment (vadz) [inline src/common/desktopenvcmn.cpp:38]: Why should we bother checking this at all? If the file doesn't exist, moving it to trash should fail anyhow.

This entire function should be just removed.

File: src/generic/desktopenv.cpp
Comment (vadz) [inline src/generic/desktopenv.cpp:30]: Failing here is probably not very useful. But it should return false, not true.

File: src/gtk/desktopenv.cpp
Comment (vadz) [inline src/gtk/desktopenv.cpp:40]: Why do you handle just this error code specifically? You should show the error message if any (i.e. if `error != NULL`) for any error code, use `wxGtkError` class from include/wx/gtk/private/error.h for it.

File: src/gtk/desktopenv.cpp
Comment (vadz) [inline src/gtk/desktopenv.cpp:46]: Does `g_file_trash()` take ownership of it on success? I.e. are you sure it doesn't need to be unref'd if it fails? I can't find documentation for this function but it would be surprising if it did this, so I think you need to use `wxGtkObject` to avoid leaks here.

File: src/osx/cocoa/desktopenv.mm
Comment (vadz) [inline src/osx/cocoa/desktopenv.mm:34]: Is it really called "recycle bin" under Mac?

File: tests/file/desktopenv.cpp
Comment (vadz) [inline tests/file/desktopenv.cpp:30]: Use `TestFile` from `tests/testfile.h` instead to ensure it's not left over.

**Review by vadz — CHANGES_REQUESTED** [review]
This is still way too complicated and half of the changes don't make sense/are unnecessary. I don't know how to explain it, but you're adding a simple function, there is no need for any base classes, multiple platform-specific headers (with exactly the same contents!) or anything else.

The only interesting question for me is whether this should be a function in core or in base. As written right now, it could work without any GUI at all too, but this would require linking with gio under Linux. Not sure if it's worth it, but OTOH there is really nothing GUI-specific in this function.

**Comment by vadz** [comment]
BTW, we already link with gio when using libsecret, which relies on it, so it probably wouldn't be a big deal to always link with it.

File: src/common/desktopenvcmn.cpp
Comment (oneeyeman1) [reply src/common/desktopenvcmn.cpp:38]: Vadim,
Why even try if the file doesn't exist? It just doesn't make sense...
And this is exactly why the base class exist - to have this check in one place and don't have it in 3 places.

I may give it a try and see if it succeeds on all 3 supported platforms without crashing the program.

File: src/gtk/desktopenv.cpp
Comment (oneeyeman1) [reply src/gtk/desktopenv.cpp:40]: Because "Not supported" is different from any other failure, and the user should know about that specific one.

Comment (vadz) [reply src/gtk/desktopenv.cpp:40]: Why is it different? Just give the associated error message, this should be clear enough and it should already say "not supported" in this case.

**Comment by PBfordev** [comment]
@oneeyeman1, as I have already commented on this feature A LOT, I was hesitant to review it again, as you could think I am trying to pick on you (I am not and I believe it or not, I do dislike to criticize). In the end, I decided to post my observations, hoping you would not mind and perhaps can find something useful in them.

I am familiar only with MSW, so all my comments are based on MSW implementation, except for one: I am not sure if `std::vector<wxString>` is the new official API replacement of `wxArrayString` or not, so I left it there; however, the downside of using it is that it does not have simple conversion to `wxArrayString`. Also, should not we `#include <vector>` somewhere before #including the platform-specific header?

### The deleted paths retrieving API
1. On Windows, the Recycle Bin can contain the same path multiple times, with differing delete dates, i.e., the content of the deleted file/folder can be different even when the path is the same. The delete date is available via `SCID_DateDeleted` property (see e.g. https://devblogs.microsoft.com/oldnewthing/20110830-00/?p=9773). I believe having the delete date for a path available is important, so the last deleted path is restored as probably expected instead of the version deleted before that. 
2. The method name is `wxDesktopEnv::GetFilesInRecycleBin()`; however, the current code allows also retrieving folders (see `SHCONTF_FOLDERS` flag in the `IShellFolder::EnumObjects()` call). I think the code should support folders and if so, perhaps it should be also able to obtain not only the path (and the deleted date) but also whether the path is a folder of a file?

### The actual MSW implementation
I find the code somehow odd, not sure if you pushed the intended version? AFAICT, the code, among else:
1.	Does not work, for example at the beginning of `wxDesktopEnv::GetFilesInRecycleBin()` you are testing the result of `SHGetKnownFolderIDList()` as if it returned true on success; however, it returns 0 (`S_OK`) when it succeeds.
2.	Contains dead code, such as retrieving the name of the Recycle Bin folder which is then never used.
3.	Uses `SHGetKnownFolderIDList()` which is available only for Vista+, while AFAIK wxWidgets 3.1 unfortunately still has to support Windows XP.
4.	Uses raw pointers instead of RAII wrappers provided by wxWidgets.
5.	Is quite difficult to follow, for example look at the length and scope level depth of `wxDesktopEnv::RestoreFromRecycleBin()`.

Instead of commenting on the perceived implementation issues individually, I attempted to rewrite the code, with focus on addressing the points listed above. You can my attempt here: https://github.com/wxWidgets/wxWidgets/commit/69bd24f05a94079445a4335729a7696b8fec0268 I hope it could be of some help.

By the way, Windows can still display a confirmation dialog when restoring the Recycle Bin item would overwrite an existing path, despite all the flags used to suppress any user interaction. Also, when the user chooses not to restore the path, the API still returns success so the code cannot tell whether the item was actually restored or the user decided not to.

Last but not least, my implementation restores the first found deleted path, which may not be the last deleted and thus not what the user expects, see the API discussion section.

FWIW, here is a simple minimal application to test enumerating the Recycle Bin and restoring an item inside it: https://gist.github.com/PBfordev/7a4865bf9ec7be14e5c716c5d81c5385

### Conclusion
I believe that the current implementation is not optimal and could be improved. But before doing anything, I would first wait for the confirmation that the maintainers are willing to accept a feature for retrieving and restoring deleted paths, as AFAIK this goes beyond the originally accepted function for moving paths to the trash.

**Comment by oneeyeman1** [comment]
@PBfordev ,
> I still do not understand the API, i.e., why have a class with a single, static, method. I guess this is supposed to be extended in the future.

It is possible to have a function here to create a shortct on the Desktop. That new function will belong to this class.
This is just one example.

> Even if so, why not submit it complete so we have a big picture available?

Because my goad is to implement "Recycle Bin" functionality.
Also the old ticket contained some stuff that could be implemented in this class.

> Nevertheless, I added a bunch of (mostly) nitpick comments.

Thank you.

**Pull request closed without merging**
